library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/async/RecordedDataQueueHandler.kt

```diff
@@ -0,0 +1,175 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.async
+
+import androidx.annotation.MainThread
+import androidx.annotation.WorkerThread
+import com.datadog.android.sessionreplay.internal.processor.RecordedDataProcessor
+import com.datadog.android.sessionreplay.internal.processor.RumContextDataHandler
+import com.datadog.android.sessionreplay.internal.recorder.SystemInformation
+import com.datadog.android.sessionreplay.internal.utils.TimeProvider
+import com.datadog.android.sessionreplay.model.MobileSegment
+import java.lang.ClassCastException
+import java.lang.NullPointerException
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.RejectedExecutionException
+
+/**
+ * Responsible for storing the Snapshot and Interaction events in a queue.
+ * Allows for asynchronous enrichment, which still preserving the event order.
+ * The items are added to the queue from the main thread and processed on a background thread.
+ */
+internal class RecordedDataQueueHandler(
+    private val processor: RecordedDataProcessor,
+    private val rumContextDataHandler: RumContextDataHandler,
+    private val executorService: ExecutorService,
+    private val timeProvider: TimeProvider
+) {
+
+    // region internal
+    internal val recordedDataQueue = ConcurrentLinkedQueue<RecordedDataQueueItem>()
+
+    @MainThread
+    internal fun addTouchEventItem(
+        pointerInteractions: List<MobileSegment.MobileRecord>
+    ): TouchEventRecordedDataQueueItem? {
+        val rumContextData = rumContextDataHandler.createRumContextData()
+            ?: return null
+
+        val item = TouchEventRecordedDataQueueItem(
+            timestamp = rumContextData.timestamp,
+            prevRumContext = rumContextData.prevRumContext,
+            newRumContext = rumContextData.newRumContext,
+            touchData = pointerInteractions
+        )
+
+        insertIntoRecordedDataQueue(item)
+
+        return item
+    }
+
+    @MainThread
+    internal fun addSnapshotItem(
+        systemInformation: SystemInformation
+    ): SnapshotRecordedDataQueueItem? {
+        val rumContextData = rumContextDataHandler.createRumContextData()
+            ?: return null
+
+        val item = SnapshotRecordedDataQueueItem(
+            timestamp = rumContextData.timestamp,
+            prevRumContext = rumContextData.prevRumContext,
+            newRumContext = rumContextData.newRumContext,
```

Comment:
minor: maybe it could leverage convenience constructor that just takes rumContextData and systemInformation as a parameters